kernels/volk/volk_32fc_index_max_avx2.tmpl

```diff
@@ -0,0 +1,131 @@
+__m256i max_indices = _mm256_setzero_si256();
+
+/*
+ * The loop below vectorizes the inner loop of the following code:
+ *
+ * float max_values[8] = {0.f};
+ * unsigned max_indices[8] = {0};
+ * unsigned current_indices[8] = {0, 1, 2, 3, 4, 5, 6, 7};
+ * for (unsigned i = 0; i < num_points / 8; ++i) {
+ *     for (unsigned j = 0; j < 8; ++j) {
+ *         float abs_squared = real(src0) * real(src0) + imag(src0) * imag(src1)
+ *         bool compare = abs_squared > max_values[j];
+ *         max_values[j] = compare ? abs_squared : max_values[j];
+ *         max_indices[j] = compare ? current_indices[j] > max_indices[j]
+ *         current_indices[j] += 8; // update for next outer loop iteration
+ *         ++src0;
+ *     }
+ * }
+ */
+for (unsigned i = 0; i < num_points / 8u; ++i) {
+    /* load 16 floats (8 complex numbers) and square them */
+#if ALIGNED == 0
+    __m256 in0 = _mm256_loadu_ps((float*)src0);
+    __m256 in1 = _mm256_loadu_ps((float*)(src0 + 4));
+#elif ALIGNED == 1
+    __m256 in0 = _mm256_load_ps((float*)src0);
+    __m256 in1 = _mm256_load_ps((float*)(src0 + 4));
+#else
+#error "Invalid value for ALIGNED"
+#endif /* ALIGNED */
+    in0 = _mm256_mul_ps(in0, in0);
+    in1 = _mm256_mul_ps(in1, in1);
+
+    /*
+     * Given the vectors a = (a_7, a_6, …, a_1, a_0) and b = (b_7, b_6, …, b_1, b_0)
+     * hadd_ps(a, b) computes
+     * (b_7 + b_6,
+     *  b_5 + b_4,
+     *  ---------
+     *  a_7 + b_6,
+     *  a_5 + a_4,
+     *  ---------
+     *  b_3 + b_2,
+     *  b_1 + b_0,
+     *  ---------
+     *  a_3 + a_2,
+     *  a_1 + a_0).
+     * The result is the squared absolute value of complex numbers at index
+     * offsets (7, 6, 3, 2, 5, 4, 1, 0).  This is the reason for the odd
+     * order of elements when current_indices is initialised.
+     */
+    __m256 abs_squared = _mm256_hadd_ps(in0, in1);
+
+    /*
+     * Compare the recently computed squared absolute values with the
+     * previously determined maximum values. cmp_ps(a, b) determines
+     * a > b ? 0xFFFFFFFF for each element in the vectors =>
+     * compare_mask = abs_squared > max_values ? 0xFFFFFFFF : 0
+     */
+    __m256 compare_mask = _mm256_cmp_ps(abs_squared, max_values, _CMP_GT_OS);
+    /*
+     * Select the maximum of the two by max() or blending (see below).
+     */
+#if VARIANT == 0
+    max_values = _mm256_blendv_ps(max_values, abs_squared, compare_mask);
+#elif VARIANT == 1
+    max_values = _mm256_max_ps(abs_squared, max_values);
+#else
+#error "Invalid value for VARIANT"
+#endif
+    /*
+     * Updates indices: blendv_ps(a, b, mask) determines mask ? b : a for
+     * each element in the vectors =>
+     * max_indices = compare_mask ? current_indices : max_indices
+     *
+     * Note: The casting of data types is required to make the compiler happy
+     * and does not change values.
+     */
+    max_indices = (__m256i)_mm256_blendv_ps(
+        (__m256)max_indices, (__m256)current_indices, compare_mask);
```

Comment:
At least MSVC doesn't understand these typecasts. I assume the correct call would be `_mm256_castps_si256` and the inverse.

Reply:
Fixed